### DISCUSSION TITLE
Converting Synchronization Primitives in C/C++ to P

### DISCUSSION THREAD

**Discussion by anupgithub** [p-org/P#450] [Q&A] [answered]
For larger and faster adoption of P Programming, tutorials on following topics would be very helpful.

1. Single Node : 
    How do we convert small programs which use mutex, semaphore, condition_variables into P. Like taking an example in c++ and then converting to P.
2.  Multi Node:
    Converting programs which use ZooKeeper Synchronization primitives to P. Or any open source distributed lock service examples.

I have been struggling to get my folks to use P because everyone is trying to map in their minds these primitives to P and we don't know if we have a good conversion documentations/videos.

**Comment by ankushdesai** [comment]
Hi Anup, Thanks for these questions.

Here is a simple example that models lock-based shared object and read-write lock based shared objects in P: https://github.com/p-org/P/blob/master/Tutorial/Common/SharedMemory/PTst/TestRWLBSharedObject.p

RWLock: https://github.com/p-org/P/blob/master/Tutorial/Common/SharedMemory/PSrc/RWLBSharedObject.p

The way to think about it is that one can model shared memory using message passing, all synchronization (using locks) that happens in shared memory programs can be mimicked using shared memory underneath. So if you look at: https://github.com/p-org/P/blob/master/Tutorial/Common/SharedMemory/PTst/TestRWLBSharedObject.p. The Acquire functions under the hood perform message passing.

I agree that a better example would help here and I will definitely add it to the tutorials.

Thanks! Please let me know your comments.

**Reply by anupgithub** [reply, marked as answer]
Thanks Ankush! We will go through the example you mentioned. 

Basically we need "dumb" synchronization examples to learn from. Take any concurreny book (c++/Java, c), converting the first 10 **solved**  concurrent problems from it to P would be a great tutorial. 

I also had the following questions:


1#  What's a atomic step in P programming? What's the smallest default atomic? 

Like if X is a integer, would the following be considered atomic? 
X = X + 1

From example in pluscal, every step is atomic. Step is defined by labels. Following is a note from lamport. 

"PlusCal is an algorithm language. Execution of an algorithm consists of a sequence of steps. An algorithm’s computational complexity is the number of steps it takes to compute the result; defining a concurrent algorithm requires specifying what constitutes a single (**atomic**) step."

2# You mentioned shared memory is implemented as messaging passing. Can you explain more about it? 

For general public this is way above comprehension. Would following be a good example?   " when CPU wants to update a memory location to some value it sends a _message_ to the memory controller and the memory controller serializes all writes from all CPUs."  Also what does a global variable mean in this construct? 



Nevertheless, thanks for taking the time Ankush. You have been a great help. If my questions burdens you please let the community help on that. Don't want to burden you too much. I wish software industry appreciated this field and funded an army of engineers/researchers for formal verification so that this becomes mainstream like other high level languages. Hopefully some day! (fingers crossed) 

-Anup
